src/main/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGenerator.java

```diff
@@ -162,14 +161,72 @@ private static DetailAST findChildWithIdent(DetailAST root) {
      * @return full xpath query for given ast element
      */
     private static String generateXpathQuery(DetailAST ast) {
-        String xpathQuery = getXpathQuery(null, ast);
-        if (!isUniqueAst(ast)) {
-            final DetailAST child = findChildWithIdent(ast);
-            if (child != null) {
-                xpathQuery += "[." + getXpathQuery(ast, child) + ']';
+        final StringBuilder xpathQueryBuilder = new StringBuilder(getXpathQuery(null, ast));
+        if (!isXpathQueryForNodeIsAccurateEnough(ast)) {
+            xpathQueryBuilder.append('[').append(findPositionAmongSiblings(ast)).append(']');
+        }
+        return xpathQueryBuilder.toString();
+    }
+
+    /**
+     * Finds position of the ast element among siblings.
+     * @param ast {@code DetailAST} ast element
+     * @return position of the ast element
+     */
+    private static int findPositionAmongSiblings(DetailAST ast) {
+        DetailAST cur = ast;
+        int pos = 0;
+        while (cur != null) {
+            if (cur.getType() == ast.getType()) {
+                pos++;
             }
+            cur = cur.getPreviousSibling();
         }
-        return xpathQuery;
+        return pos;
+    }
+
+    /**
+     * Checks if ast element has all requirements to have unique xpath query.
+     * @param ast {@code DetailAST} ast element
+     * @return true if ast element will have unique xpath query, false otherwise
+     */
+    private static boolean isXpathQueryForNodeIsAccurateEnough(DetailAST ast) {
+        return !hasAtLeastOneSiblingWithSameTokenType(ast)
+                || XpathUtil.supportsValueAttribute(ast)
+                || findChildWithValueAttribute(ast) != null;
+    }
+
+    /**
+     * Checks if the given ast element has unique {@code TokenTypes} among siblings.
+     * @param ast {@code DetailAST} ast element
+     * @return if the given ast element has unique {@code TokenTypes} among siblings
+     */
+    private static boolean hasAtLeastOneSiblingWithSameTokenType(DetailAST ast) {
+        boolean result = false;
+        if (ast.getParent() == null) {
```

Comment:
removed conditional - replaced equality check with true → SURVIVED

Reply:
@rnveach, why CI is not failing ?

Reply:
@romani These packages aren't set up for these pitest mutators at the moment. I added the mutators locally and ran the report. This specific mutator I am still trying to work through in https://github.com/checkstyle/checkstyle/issues/6320 .

Reply:
If you want to do that, I am fine with it.

Reply:
At least it will be clear for contributors, what is expected and what to run to verify fix